wrangler-core/src/main/java/io/cdap/functions/DateAndTime.java

```diff
@@ -0,0 +1,663 @@
+   * @param microseconds component of {@link LocalTime}
+   * @return a instance of {@link LocalTime}
+   */
+  public static LocalTime TimeFromComponents(Integer hours, Integer mins, Integer seconds, Integer microseconds) {
+    return hours == null || mins == null || seconds == null || microseconds == null ? null :
+             LocalTime.of(hours, mins, seconds, microseconds * 1000);
+  }
+
+  /**
+   * Returns the time given the number of seconds since midnight. If seconds is null, midnight will be returned.
+   *
+   * @param seconds since midnight.
+   * @return time since midnight.
+   */
+  public static LocalTime TimeFromMidnightSeconds(Integer seconds) {
+    return LocalTime.of(0,0,0).plusSeconds(seconds == null ? 0 : seconds);
+  }
+
+  /**
+   * Returns the time, with offsets applied from the base time.
+   * hour offset, minute offset, and second offset each given as separate values.
+   * The seconds offset can include partial seconds.
+   * If time is null, null will be returned. If the offset is null, it will be treated as 0.
+   *
+   * @param time baseline {@link LocalTime}
+   * @param hour offset to added to baseline.
+   * @param min offset to added to baseline.
+   * @param seconds offset to added to baseline.
+   * @return hour, min and second added to baseline.
+   */
+  public static LocalTime TimeOffsetByComponents(LocalTime time, Integer hour, Integer min, Integer seconds) {
+    return time == null ? null : time.plusHours(hour == null ? 0 : hour)
+                                   .plusMinutes(min == null ? 0 : min)
+                                   .plusSeconds(seconds == null ? 0 : seconds);
+  }
+
+  /**
+   * Returns a timestamp from the given date and time. If any input is null, null will be returned.
+   *
+   * @param date a instance of {@link LocalDate}
+   * @param time a instance of {@link LocalTime}
+   * @return a instance of {@link LocalDateTime}
+   */
+  public static LocalDateTime TimestampFromDateTime(LocalDate date, LocalTime time) {
+    return date == null || time == null ? null : LocalDateTime.of(date, time);
+  }
+
+  /**
+   * Returns a timestamp that is derived from the number of seconds from the base timestamp object.
+   * If time is null, null will be returned. If seconds is null, time will be returned.
+   *
+   * @param seconds to be added to base timestamp.
+   * @param time base timestamp.
+   * @return timestamp with seconds added to base time.
+   */
+  public static LocalDateTime TimestampFromSecondsSince(Integer seconds, LocalDateTime time) {
+    return time == null ? null : time.plusSeconds(seconds == null ? 0 : seconds);
+  }
+
+  /**
+   * Returns a timestamp from the given epoch time specified in seconds. If timestamp is null, null will be returned.
+   *
+   * @param epochTimestamp specified in seconds.
+   * @return a instance {@link LocalDateTime}
+   */
+  public static LocalDateTime TimestampFromEpoch(Long epochTimestamp) {
```

Comment:
any timestamp seems like it should be using ZonedDateTime, as it represents an exact point in time. That is what StructuredRecord uses too.

Reply:
The DataStage date and time functions basically do not operate on a ZonedDateTime, here is their doc: https://www.ibm.com/support/knowledgecenter/SSZJPZ_11.7.0/com.ibm.swg.im.iis.ds.parjob.dev.doc/topics/r_deeref_Date_and_Time_Functions.html.
We have also heard about use cases about these functions used in LocalDateTime. I feel these functions are more like the Datetime type in Bigquery.

Reply:
This is still weird to me, as it would be returning a LocalDateTime assuming it is UTC, but losing the fact that it is UTC. I don't see a TimestampFromEpoch function in the link you gave.

I think it makes more sense to have this return a ZonedDateTime, and have another `ToDateTime(ZonedDateTime timestamp, String timeZone)` function used to convert a real timestamp to a date time.

Reply:
Actually, it's called TimestampFromTimet in the doc you shared. Can have a DateTimeFromEpoch(), but it should be documented that it will be a DateTime in UTC.